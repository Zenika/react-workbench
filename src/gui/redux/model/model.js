import { RECEIVED_DOCGEN } from '../docgen'
import { docgenToModel, convertFromGuiValue } from '../utils/docgen.js'

export const UPDATE_VALUE = 'UPDATE_VALUE'

export default (state = {}, { type, payload }) => {
  switch (type) {
    case RECEIVED_DOCGEN:
      return docgenToModel(payload)
    case UPDATE_VALUE: {
      return { ...state, [payload.name]: convertFromGuiValue(payload.value, payload.type) }
    }
    default:
      return state
  }
}
