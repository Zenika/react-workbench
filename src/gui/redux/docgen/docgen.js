import factory from 'trampss-redux-factory'

export default factory({ name: 'docgen', type: 'uniq' })
