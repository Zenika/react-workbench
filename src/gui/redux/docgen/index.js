export { default } from './docgen'
export * from './docgen.selectors'
export * from './docgen.api'
