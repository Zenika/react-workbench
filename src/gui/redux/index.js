export * from './store'
export { default } from './store'
