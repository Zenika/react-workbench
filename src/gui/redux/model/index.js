export { default } from './model'
export * from './model.selectors'
export * from './model.actions'
