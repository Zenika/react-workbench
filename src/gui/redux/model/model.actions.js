export const SET_MODEL = 'SET_MODEL'
export const setModel = model => ({ type: SET_MODEL, payload: model })

export const SET_VALUE = 'SET_VALUE'
export const setValue = (name, value) => ({ type: SET_VALUE, payload: { name, value } })

const getDocgenValue = (value, type) => {
  switch (type) {
    case 'bool':
      return value === 'true'
    case 'object':
      return eval(`Object(${value})`) // eslint-disable-line no-eval
    case 'symbol':
      return value
    default:
      return eval(value) // eslint-disable-line no-eval
  }
}

export const fetchModel = () => (dispatch) => {
  fetch('/api/docgen').then(response => response.json()).then((json) => {
    if (json && json.length > 0) {
      dispatch(
        setModel(
          Object.keys(json[0].props || {}).reduce((obj, key) => {
            const prop = json[0].props[key]
            const type = prop.type && prop.type.name
            const value = prop.defaultValue && prop.defaultValue.value

            return {
              ...obj,
              [key]: {
                value: getDocgenValue(value, type),
                type,
              },
            }
          }, {})
        )
      )
    }
  })
}
