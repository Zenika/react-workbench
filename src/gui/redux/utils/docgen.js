const convertDocgenValue = (value, type) => {
  switch (type) {
    case 'bool':
      return value === 'true'
    case 'object':
      return eval(`Object(${value})`) // eslint-disable-line no-eval
    case 'symbol':
      return value
    default:
      return eval(value) // eslint-disable-line no-eval
  }
}

const docgenToModel = (docgen) => {
  return Object.keys(docgen.props || {}).reduce((obj, key) => {
    const prop = docgen.props[key]
    const type = prop.type && prop.type.name
    const value = prop.defaultValue && prop.defaultValue.value
    return { ...obj, [key]: convertDocgenValue(value, type) }
  }, {})
}

const convertToGuiValue = (value, type) => {
  switch (type) {
    case 'array':
    case 'object':
      return JSON.stringify(value)
    case 'func':
      return eval(value) // eslint-disable-line no-eval
    default:
      return value
  }
}

const convertFromGuiValue = (value, type) => {
  switch (type) {
    case 'bool':
      return !!value
    case 'object':
      return eval(`Object(${value})`) // eslint-disable-line no-eval
    case 'array':
    case 'func':
      return eval(value) // eslint-disable-line no-eval
    default:
      return value
  }
}

export { convertDocgenValue, docgenToModel, convertToGuiValue, convertFromGuiValue }
