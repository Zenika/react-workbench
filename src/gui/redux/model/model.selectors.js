import { createSelector } from 'reselect'

import model from './model'
import { convertToGuiValue } from '../utils/docgen'

const getModelData = state => model.get()(state)

const getModelArray = state => model.getAsArray(state)

const getComponentValue = (name, type) =>
  createSelector(getModelData, data => convertToGuiValue(data[name].value, type))

const getComponentProps = createSelector(getModelArray, array =>
  array.map(p => ({ [p.name]: p.value })).reduce((acc, curr) => ({ ...acc, ...curr }), {})
)

export { getComponentProps, getComponentValue }
