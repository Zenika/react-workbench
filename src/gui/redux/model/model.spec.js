/* eslint-env jest */
import { createStore, combineReducers } from 'redux'
import docgen from '../docgen'
import model from './model'

describe('gui/redux/model', () => {
  let store

  beforeEach(() => {
    store = createStore(
      combineReducers({
        model,
        docgen,
      })
    )
  })

  it('should transform SET_DOCGEN to SET_MODEL action', () => {
    // dispatch docgen set action
    store.dispatch(
      docgen.set({
        props: {
          foo: {
            type: { name: 'string' },
            defaultValue: { value: "'foo'" },
          },
        },
      })
    )

    // check if model state is updated
    expect(model.isInitialized(store.getState())).toBe(true)
    expect(model.getKeys(store.getState()).includes('foo')).toBe(true)
  })

  it('should convert gui values to model values', () => {
    // dispatch docgen set action
    store.dispatch(
      docgen.set({
        props: {
          foo: {
            type: { name: 'string' },
            defaultValue: { value: "'foo'" },
          },
        },
      })
    )
    // dispatch docgen model action
    store.dispatch(model.update({ name: 'foo', value: 'bar', type: 'string' }))

    // check if model is updated
    expect(model.get('foo')(store.getState()).value).toBe('bar')
  })
})
