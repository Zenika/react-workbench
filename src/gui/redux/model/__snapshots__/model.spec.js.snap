// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gui/redux/model reducer should convert gui values to model values 1`] = `
Object {
  "array": Array [
    Object {
      "name": "foo",
      "type": "string",
      "value": "bar",
    },
  ],
  "data": Object {
    "foo": Object {
      "name": "foo",
      "type": "string",
      "value": "bar",
    },
  },
  "initialized": true,
  "keys": Array [
    "foo",
  ],
}
`;

exports[`gui/redux/model reducer should transform SET_DOCGEN to SET_MODEL action 1`] = `
Object {
  "array": Array [
    Object {
      "name": "foo",
      "value": "foo",
    },
  ],
  "data": Object {
    "foo": Object {
      "name": "foo",
      "value": "foo",
    },
  },
  "initialized": true,
  "keys": Array [
    "foo",
  ],
}
`;
